### PR TITLE
Move EKF buffer handling to common classes

### PR DESCRIPTION
**Pull request by tridge**
This moves EKF buffer handling out of the template classes in each directory into common classes. This saves about 6k of flash
This was test flown today with the following log:
http://uav.tridgell.net/DartXL/2020-11-15/00000016.BIN
the log had both EK2 and EK3 enabled. Both EKF instances behaved normally
@bugobliterator I've tagged you as I believe you were the original author of this buffer code
Note that pop_oldest_element() was renamed to get_oldest_element() as it doesn't actually remove an element from the buffer

### CODE REVIEW

**Comment by tridge** [comment]
@rmackay9 this PR rebases with a one line trivial header fix on your sources PR

File: libraries/AP_NavEKF3/AP_NavEKF3_core.h [hunk 1]
```diff
@@ -550,87 +550,75 @@ class NavEKF3_core : public NavEKF_core_common
...
-    struct yaw_elements {
+    struct yaw_elements : EKF_buffer_element_t {
         float       yawAng;         // yaw angle measurement (rad)
         float       yawAngErr;      // yaw angle 1SD measurement accuracy (rad)
-        uint32_t    time_ms;        // measurement timestamp (msec)
         uint8_t     type;           // type specifiying Euler rotation order used, 1 = 312 (ZXY), 2 = 321 (ZYX)
     };
 
-    struct ext_nav_elements {
+    struct ext_nav_elements : EKF_buffer_element_t {
```
Comment (peterbarker) [inline libraries/AP_NavEKF3/AP_NavEKF3_core.h:612]:
```suggestion
    struct ext_nav_elements : EKF_obs_element {
```

File: libraries/AP_NavEKF2/AP_NavEKF2_core.h [hunk 2]
```diff
@@ -841,13 +832,13 @@ class NavEKF2_core : public NavEKF_core_common
...
     float gpsNoiseScaler;           // Used to scale the  GPS measurement noise and consistency gates to compensate for operation with small satellite counts
     Matrix24 P;                     // covariance matrix
-    imu_ring_buffer_t<imu_elements> storedIMU;      // IMU data buffer
-    obs_ring_buffer_t<gps_elements> storedGPS;      // GPS data buffer
-    obs_ring_buffer_t<mag_elements> storedMag;      // Magnetometer data buffer
-    obs_ring_buffer_t<baro_elements> storedBaro;    // Baro data buffer
-    obs_ring_buffer_t<tas_elements> storedTAS;      // TAS data buffer
-    obs_ring_buffer_t<range_elements> storedRange;  // Range finder data buffer
-    imu_ring_buffer_t<output_elements> storedOutput;// output state buffer
+    EKF_IMU_buffer_t<imu_elements> storedIMU;      // IMU data buffer
```
Comment (rmackay9) [inline libraries/AP_NavEKF2/AP_NavEKF2_core.h:835]: nitpick: minor alignment issue on the comment..

**Comment by tridge** [comment]
call comments:
use EKF_obs_buffer
try : based struct init

File: libraries/AP_NavEKF/EKF_Buffer.cpp [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    // Advance head to next available index
+    _head = (_head+1) % _size;
+    // New data is written at the head
+    memcpy(get_offset(_head), element, elsize);
+    _new_data = true;
+}
+
+
+// writes the same data to all elements in the ring buffer
+void ekf_ring_buffer::reset_history(const void *element)
```
Comment (tridge) [inline libraries/AP_NavEKF/EKF_Buffer.cpp:114]: can we remove this?

File: libraries/AP_NavEKF/EKF_Buffer.cpp [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+        if (time_ms(tail) != 0 && time_ms(tail) <= sample_time) {
+            // if head is equal to tail just check if the data is unused and within time horizon window
+            if (((sample_time - time_ms(tail)) < 100)) {
+                bestIndex = tail;
+                success = true;
+                _new_data = false;
+            }
+        }
+    } else {
+        while(_head != tail) {
```
Comment (peterbarker) [inline libraries/AP_NavEKF/EKF_Buffer.cpp:69]: I wonder if we miss checking the last element in the list.

File: libraries/AP_NavEKF/EKF_Buffer.cpp [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+// initialise buffer, returns false when allocation has failed
+bool ekf_imu_buffer::init(uint32_t size)
+{
+    buffer = calloc(size, elsize);
+    if (buffer == nullptr) {
+        return false;
+    }
+    _size = size;
+    _youngest = 0;
+    _oldest = 0;
```
Comment (peterbarker) [inline libraries/AP_NavEKF/EKF_Buffer.cpp:150]: Any reason to think these won't be zero anyway?

File: libraries/AP_NavEKF2/AP_NavEKF2_core.h (hunk 2 above)
Comment (tridge) [reply libraries/AP_NavEKF2/AP_NavEKF2_core.h:835]: anti-nitpick, it is aligned (its a viewer issue in github)

File: libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp [hunk 6]
```diff
@@ -1040,12 +1040,13 @@ void NavEKF3_core::writeExtNavVelData(const Vector3f &vel, float err, uint32_t t
...
-        err,
-        timeStamp_ms,
-        false
-    };
+
+    ext_nav_vel_elements extNavVelNew;
+    extNavVelNew.time_ms = timeStamp_ms;
+    extNavVelNew.vel = vel;
+    extNavVelNew.err = err;
+    extNavVelNew.corrected = false;
```
Comment (peterbarker) [inline libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:1048]: Should be able to use designator initialisation... hopefully...

**Review by peterbarker — CHANGES_REQUESTED** [review]

File: libraries/AP_NavEKF/EKF_Buffer.cpp [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+bool ekf_ring_buffer::init(uint8_t size)
+{
+    buffer = calloc(size, elsize);
+    if (buffer == nullptr) {
+        return false;
+    }
+    _size = size;
+    _head = 0;
+    _tail = 0;
+    _new_data = false;
```
[the remark this answers is not in the document]
Comment (tridge) [reply libraries/AP_NavEKF/EKF_Buffer.cpp:24]: no

File: libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp (hunk 6 above)
Comment (tridge) [reply libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp:1048]: nope, doesn't work

**Review by peterbarker — APPROVED** [review]

**Pull request merged**